### DISCUSSION TITLE
Correct way to load dynamic gcode within the app

### DISCUSSION THREAD

**Discussion by alastaira** [arkypita/LaserGRBL#1428] [General]
Hi!

I've written myself a couple of tools to automate various GRBL tasks (e.g. to create power/speed test cards, insert text, parallel line tests etc.) and, as they were written in C# anyway, I thought it made sense to integrate them into LaserGRBL.
I've got them working, but to do so I've had to hack GrblCore.cs a bit to be enable to load GRBL generated at runtime, since all the functions I could find loaded data from physical (NC/SVG/PNG/JPG) files stored on disk. So I created a new method which inserted a string of arbitrary gcode into the "list" field of Core. like this:
```
public void LoadGCodeString(string gCode, bool append) {
            RiseOnFileLoading("Commands");
            long start = Tools.HiResTimer.TotalMilliseconds;

            if (!append)
                list.Clear();

            // Copy the string into a memory stream for ease of access
            var stream = new System.IO.MemoryStream();
            var writer = new System.IO.StreamWriter(stream);
            writer.Write(gCode);
            writer.Flush();
            stream.Position = 0;

            using (System.IO.StreamReader sr = new System.IO.StreamReader(stream)) {
                string line = null;
                while ((line = sr.ReadLine()) != null)
                    if ((line = line.Trim()).Length > 0) {
                        GrblCommand cmd = new GrblCommand(line);
                        if (!cmd.IsEmpty)
                            list.Add(cmd);
                    }
            }

            Analyze();
            long elapsed = Tools.HiResTimer.TotalMilliseconds - start;

            RiseOnFileLoaded("Commands", elapsed);
        }
```

Now, this works great, but I've just spotted the "Enqueue" method instead - is that the correct approach I should have used for this? I've just tried replacing my code to submit each GRBL instruction via the Enqueue() method instead and, while it didn't generate any compile or runtime errors, nor did it seem to update the preview window with any of the commands I added?
Thanks.

**Comment by arkypita** [comment]
You took a cue from the `public void LoadFile (string filename, bool append)` method, which is the one to load a .nc file, and you made a copy of it, modified to load a string content. This is a correct procedure.

Enqueue is a much more low level command.

Since the operation you are doing is equivalent to loading a new content/file, I suggest you enter a test like `if (!CanLoadNewFile) return;` in an external call (i.e in GrblCore) to prevents your tools from resetting the contents of "list" and enqueue new commands when it is not a good time to do so.

**Comment by arkypita** [comment]
Unfortunately the LaserGRBL code was written "quick and dirty" without a project at the base inserting different functionalities without unifying (for example the generation of the gcode for the SVG and for the vectorized raster follows two radically different paths) and without worrying too much about making it modular and expandable

Unfortunately this makes it difficult to contribute other people's code, as well as the development of some functions that would be very useful (for example layers, SVG filling, DXF support) is uncomfortable.

Sooner or later I'll have to think about a rewrite of the code, such as to allow easier maintenance and development.

In the meanwhile If you want to contribute with some functions and have any questions, I will try to answer you as much as possible.

In any case, I suggest you use github to synchronize your working copy with the arkypita/LaserGRBL trunk and also to publish and make your pull requests available for possible distribution in the official LaserGRBL version.

**Comment by arkypita** [comment]
P.S: [someone already try to do something similar](https://github.com/arkypita/LaserGRBL/pulls) but I never merged/accepted their pull because they didn't seem well integrated.

Furthermore, conceptually I have always rejected this kind of development because the basic choice below lasergrbl is not to make it a graphic editor. 

If I add the function that puts the text, then people wants to add a circle as well, and then also a star, and then he wants to move them, and then he wants to align them. For all these things there are a thousand other programs, vector and raster, that do everything you could wish for.

**Reply by alastaira** [reply]
Thanks for the reply! Conceptually, I actually completely agree with you - I use Inkscape for all my design work, and typically just import the file into LaserGRBL and burn it completely unchanged - I think it is much better to keep the interface clean and simple, and  only try to do one job and do it well :)

My motivation is *not* to create a generic shape/text tool - as you said, there would be a ton of fiddly little details that people would keep wanting to add - rather, I wanted to have the ability to create various "test" files - e.g. 
- a set of parallel lines to test for axis alignment/wobble
- a raster fill test that compares darkness of different power/speed combinations
- a cut test to work out the optimum no of passes/speed to make a clean cut through a piece etc.
- a text test that confirms legibility of a font different sizes

Of course, I can create these files separately in a standalone app, but the advantage of creating them within LaserGRBL itself was that I could read and adjust the tests based on the saved GRBL configuration values, and also quickly iterate and send different variations without needing a separate tool.

p.s. I almost exclusively work with vector files. I currently use an Inkscape extension script I wrote in Python to separate and save out separate SVG files for cut/outline/fill etc. and then File->Append them into LaserGRBL with the appropriate speed/power settings for each. Would be happy to help try to contribute to moving that procedure into LaserGRBL instead but, like you said, it's probably dependent on there being the structure to support multi-layer projects first.

**Comment by alastaira** [comment]
I've forked the repository and uploaded my draft changes to https://github.com/playfultechnology/LaserGRBL - feel free to review if you'd like, but I won't submit a PR because I understand your reasons for wanting to keep LaserGRBL pure and slimline!
